Fixes for weights with linear fitter

saimn opened the pull request:
Fix #7035.
I still find the `modeling` code quite confusing between the data axes, model sets, parameters, so I hope that I did it right for all cases (and tests seem to show that it works).

jehturner commented:
Thanks, @saimn! I may be wrong, but I think this can be simplified significantly. Let me check...

jehturner commented:
No, that idea was just a misunderstanding of how you've done it and why -- but could we maybe combine the separate loops for `masked` & `weights`? I could work on that if you want (I also had one or two remaining uncertainties about the 2D weights behaviour that I could check at the same time).

I think it's good to have several tests where the weights aren't all 1, to make sure the results are correct and not just that the code runs. I see you do have one test that uses real weights for a single 2D model and also one for a 1D model set, which covers the main cases of looping & not looping. I wonder about adding some real weights for a 2D model set, to make sure the algebra is still correct in that case, but this is already pretty good and it's great that you're covering all these cases.

saimn commented:
> could we maybe combine the separate loops for masked & weights? I could work on that if you want

Sure, another pair of eyes would be useful ;).

> I think it's good to have several tests where the weights aren't all 1

Indeed. I needed tests to cover quickly all the different kind of inputs, so I basically copied the existing tests and added weights, but using something else than 1 requires to spend more time understanding each test case. We should certainly do that, if the algorithmic part is on the good path.

saimn commented on astropy/modeling/fitting.py:
```diff
@@ -446,8 +461,18 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
+                        weights = np.rollaxis(z, model_axis, z.ndim)
+                        weights = weights.reshape(-1, weights.shape[-1])
+                    else:
+                        weights = weights.flatten()
```
This I'm not really sure about, because it doesn't match the else clause for the data just above, but I didn't really understand this case.

jehturner replied:
Yes, that's one of the main things I didn't understand. It doesn't look right but I'll try to figure it out.

jehturner replied:
I believe I've figured out what to do about this, but haven't done a lot more after a somewhat interrupted day (mostly by screaming kids, which is a bit of a nightmare when trying to keep all these different cases in one's head...). The current `else` here is handling multiple cases that need separating and there are one or two other changes needed to fix it, including an additional test or two for a case that currently fails.

jehturner reviewed (commented):
I haven't actually finished reviewing this but I'll do "finish your review" because I think you won't see my couple of line comments otherwise? How do you want any changes I make? I would think a PR on your PR (like Curtis's) is best, since they'll probably be too complicated for github suggestions. At least this should be easier now we can check further changes against your tests...

saimn commented:
@jehturner - Yes, a PR on my PR would work :+1:

pllim commented:
> PR on my PR

Random thought: One of these days, we need to see how many PR to PR we can chain for a single PR. 😄

jehturner commented:
Maybe we could try a circular one?

jehturner commented on astropy/modeling/fitting.py:
```diff
@@ -446,8 +461,18 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
+                if weights is not None:
+                    # Same for weights
+                    if weights.ndim > 2:
+                        weights = np.rollaxis(z, model_axis, z.ndim)
```
Uh oh, this says `z` instead of `weights`!

saimn replied:
Oops, good catch, fixed.

saimn commented on astropy/modeling/fitting.py:
```diff
@@ -482,14 +507,20 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
-                # Don't modify in-place in case rhs was the original dependent
-                # variable array
-                rhs = rhs * weights[:, np.newaxis]
+                if weights is not None and weights.shape == rhs.shape:
```
Fixed.

jehturner commented on astropy/modeling/tests/test_fitters.py:
```diff
@@ -291,6 +326,35 @@ def test_linear_fit_model_set_fixed_parameter(self):
+    def test_linear_fit_2d_model_set_common_weight(self):
+        init_model = models.Polynomial2D(degree=2, c1_0=[1, 2], c0_1=[-0.5, 1],
+                                         n_models=2,
+                                         fixed={'c1_0': True, 'c0_1': True})
```
Just wondering whether there's any special reason for fixing the parameters here and in the other test?

nden commented:
Is this ready for review?
@jehturner You are most familiar with this code, I'll wait for your approval before I review it.

jehturner commented on astropy/modeling/fitting.py:
```diff
@@ -499,7 +543,29 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
         masked = np.any(np.ma.getmask(rhs))
 
-        if len(model_copy) == 1 or not masked:
+        if weights is not None and weights.ndim > 1 and weights.shape == rhs.shape:
```
I think you don't need the 3rd condition here (`shape`), because `weights` can only have a model set axis at this point if the `rhs` also has one (and their shapes should have been checked earlier).

saimn replied:
Ok, I removed the condition.

jehturner commented:
I think the code is correct now, @nden, except that I was still checking what happens with `scl`. It would be good if the separate loops for `weights.ndim > 1` and `len(model_copy) > 1 and masked` could be consolidated in a reasonably efficient way though; I was hoping to look at that last week but didn't get to it. I think @saimn and I are both out next week and am not sure if one of us will get to it before then (or convince ourselves it would be too messy).

jehturner commented on astropy/modeling/fitting.py:
```diff
@@ -1299,16 +1369,23 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
+            raise ValueError("x and y should have the same shape")
         farg = (x, y)
     else:
+        if not (x.shape == y.shape == z_shape):
+            raise ValueError("x, y and z should have the same shape")
```
This check that I added is currently blocking the case of a length-1 model set axis, which otherwise works, so we should perhaps change that and add a test for it. This happens because the criterion for dealing with the model set axis in `_convert_inputs` is `n_models > 1` rather than the actual array dimensionality.

jehturner replied:
In the existing code, however, providing y/z/weights with a length-1 model set axis **only works for 2D models**, not 1D ones, so this usage might not actually have been intended to work, @nden? But people could still be using it and it would seem more convenient if it does work (in both cases).

saimn replied:
How is it useful to have length-1 weights ?

jehturner replied:
Not `weights` specifically, also `y` or `z`. People might already be using this, eg. because they have a script that fits some range of image rows that can sometimes be a single row (and therefore ends up with a dimension of 1 unless they explicitly check for that case). I think we may as well make life easier by tolerating that case, rather than breaking people's stuff.

jehturner replied:
I have a fix for this.

jehturner replied:
I think that takes care of my concerns, except that it would be good to combine the loops I mentioned (but I doubt I'd be able to look at that now before the week after next).

saimn commented:
@nden @perrygreenfield - This is now ready for review, thanks to @jehturner for addressing the remaining items :)

pllim commented on astropy/modeling/tests/test_fitters.py:
```diff
@@ -873,6 +1002,134 @@ def test_fitters_with_weights():
+        assert_allclose(model.c1_0, -0.1, atol=1e-12)
+        assert_allclose(model.c0_1, 0.2, atol=1e-12)
+
+    # # The following 2 cases also happen to work, but model set fitting is
```
Why not use `pytest.mark.skip`?

saimn replied:
Not sure if we need to keep those tests, @jehturner ?

jehturner replied:
I don't think we particularly need to keep them, but if they're there they could just be uncommented when those fitters support model sets. They do also work now, but I'm not sure it's reasonable to require those cases to continue working as long as the fitters aren't advertised to support model sets (so there's no reason for there to be an extra dimension).

pllim reviewed (commented):
Any plans to address codecov comment about certain LOCs not covered? If it is not covered on purpose, can add `# pragma: no cover` directive too.

Need another rebase for the change log conflict.

p.s. When this is merge, the _weight_ is off your shoulder! 😉

saimn commented:
We (with @jehturner) are working on integrating this in our fitting code, it works well, and I just found a little issue when weights for one of the models are all 0 and the data is not a masked array. In that case the lapack routine in Numpy will crash with a rather obscure exception (`ValueError: On entry to DLASCL parameter number 4 had an illegal value`) so I added a check for this to raise a better exception. If the data is a masked array there will be a `RuntimeWarning: invalid value encountered in true_divide` but the code does not crash and just return a model with NaN parameters.

ping @nden @perrygreenfield - please review when you can :)

nden commented:
This is a substantial improvement! 
I have only one request - move the model set tests to `test_model_sets.py`.

Because of the late review the change log needs to be moved to `4.0.4`. 
And please squash commits. Thanks!

saimn commented:
@nden - I moved most of the model set tests to `test_model_sets.py`, but there were already some model set tests in `test_fitting.py`, and for some of the new model set tests we use the fixtures/class from `test_fitting.py`.
I also squashed some commits (e.g. changelog related), but I'm not sure what you want me to do here ? Do you want a single commit for this whole blob of changes ? IMO some of them deserve their own commit to be understandable (if one has to browse the history in the future) and some have comments that could be helpful as well...

saimn commented:
Ping @nden, would be super useful for us to have this in the next bugfix releases (i.e. 4.1.1).

saimn commented on astropy/modeling/tests/test_model_sets.py:
```diff
@@ -311,3 +315,262 @@ def test_fitting_shapes():
+        assert_allclose(model.c1_0, -0.1, atol=1e-12)
+        assert_allclose(model.c0_1, 0.2, atol=1e-12)
+
+    # # The following 2 cases also happen to work, but model set fitting is
```
I talked with @jehturner and I think it's better to remove them, since this is not a supported feature. If we can keep them in the history, maybe at some point we can find them again if it's useful.

saimn replied:
I pushed a commit to remove these tests, and skipped CI since this is removing only commented lines.

pllim commented:
RTD warnings are real but unrelated. Merging then. Thanks, everyone!
